### ISSUE TITLE
No ToolChain found Mac OSX

### ISSUE THREAD

**Issue by abhijit-inge** [bazelbuild/bazel#4995]
> ATTENTION! Please read and follow:
> - if this is a _question_ about how to build / test / query / deploy using Bazel, ask it on StackOverflow instead: https://stackoverflow.com/questions/tagged/bazel
> - if this is a _discussion starter_, send it to bazel-discuss@googlegroups.com
> - if this is a _bug_ or _feature request_, fill the form below as best as you can.

### Description of the problem / feature request:
Bazel build fails on Mac OS  with the following error

Analysis of target '//examples/macos/HelloWorld:HelloWorld' failed; build aborted: No toolchain found for --cpu='darwin_x86_64' --compiler='compiler'. Valid toolchains are: [
  --cpu='armeabi-v7a' --compiler='compiler' --glibc='armeabi-v7a',
  --cpu='ios_x86_64' --compiler='compiler' --glibc='ios',
  --cpu='darwin' --compiler='compiler' --glibc='macosx',
  --cpu='x64_windows' --compiler='mingw-gcc' --glibc='mingw',
  --cpu='x64_windows' --compiler='msvc-cl' --glibc='msvcrt',
]
> Replace this line with your answer.

### Feature requests: what underlying problem are you trying to solve with this feature?

> Replace this line with your answer.

### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

> Replace this line with your answer.

### What operating system are you running Bazel on?

> MAC OSX 10.12.6

### What's the output of `bazel info release`?

> release 0.11.1-homebrew

### If `bazel info release` returns "development version" or "(@non-git)", tell us how you built Bazel.

> Replace this line with your answer.

### What's the output of `git remote get-url origin ; git rev-parse master ; git rev-parse HEAD` ?

> Not a git repository.

###  Have you found anything relevant by searching the web?

>Nothing concrete so far
>
> Places to look:
> - StackOverflow: http://stackoverflow.com/questions/tagged/bazel
> - GitHub issues: https://github.com/bazelbuild/bazel/issues
> - email threads on https://groups.google.com/forum/#!forum/bazel-discuss

### Any other information, logs, or outputs that you want to share?

> Replace these lines with your answer.
>
> If the files are large, upload as attachment or provide link.

**Comment by jin** [comment]
What is `//examples/macos/HelloWorld:HelloWorld`? Do you have a reproducible project/repository that we can take a look at?

**Comment by abhijit-inge** [comment]
Hi, you can Find it here,

https://github.com/bazelbuild/rules_apple/tree/master/examples/macos

**Comment by jin** [comment]
Do you mind opening this issue on the rules_apple repository? https://github.com/bazelbuild/rules_apple/issues 

Thanks!

**Comment by abhijit-inge** [comment]
I tried all the options, passing --cpu=darwin_x86_64, --cpu=darwin and not passing any flag at all. But, unfortunately I always ran into the same issue.

**Comment by sergiocampama** [comment]
do you have Xcode installed? and the command line tools? Also, please send the full options as returned by `--announce_rc`

**Comment by RNabel** [comment]
Using `rules_apple`, can *not* reproduce (macOS 10.13.4, bazel `0.11.0-homebrew`). When setting the `--cpu` to an incorrect value the following options are listed as valid: 
```
  darwin,
  darwin_x86_64,
  k8,
  ios_x86_64,
  ios_i386,
  ios_armv7,
  ios_arm64,
  watchos_i386,
  watchos_armv7k,
  tvos_x86_64,
  tvos_arm64,
  armeabi-v7a,
```
@abhijit-inge are you using a custom `CROSSTOOL`?

While I cannot reproduce this error in `rules_apple`, I do encounter this error with our CROSSTOOL and cross-compiling for other platforms (consoles). @buchgr do you happen to be aware of a reason why `x64_windows` could override the specified option? I can open a separate issue if needed, and find a minimum repro case.

The format of the error I'm seeing on Windows is:
building with `--cpu=cpu_a --compiler=compiler_a --glibc=glibc_a` getting an error similar to:

```
No toolchain found for --cpu='x64_windows' --compiler='compiler_a' --glibc='glibc_a'. Valid toolchains are: [
<various toolchains, including x64_windows toolchains>
--cpu='cpu_a' --compiler='compiler_a' --glibc='glibc_a',
]
```
In some cases, this was fixed by specifying `--host_cpu` instead of `--cpu` which made no sense whatsoever.

**Comment by mhlopko** [comment]
I'll try to summarize, please tell me where exactly is the problem or if the problem is resolved.

When specifying C++ toolchain, you have to use --crosstool_top to specify which cc_toolchain_suite to use, and then use the triplet --cpu, --compiler, and --glibc to select single cc_toolchain from the suite (and to select the right toolchain from the CROSSTOOL). By default the host C++ toolchain is derived from the target toolchain, so if you're specifying a cc_toolchain_suite that doesn't work with your machines's cpu, you have to override host C++ toolchain using --host_crosstool_top, --host_cpu, --host_compiler, and --host_glibc.

You can specify default_toolchain in the CROSSTOOL, that will allow you to omit --compiler and --glibc for a particular toolchain.

I fully agree that the error messages are not very clear. We're not investing much into making this area of the code better, since @katre is doing great progress with his platforms work and soon (Q3?) the C++ toolchain selection mechanism will be completely different (and much more sane).

**Comment by RNabel** [comment]
First off, using `default_toolchain` was the solution for us. 

The "problem" is that it is not clear _how_ a toolchain is selected. I'd love to hear more about how a "C++ toolchain is derived from the target toolchain". If I specify `--cpu` of a console, say a PS4, I expect it to be compiled for that target (provided the correct custom CROSSTOOL has been specified). So my question is: why, if there is a valid combination of, say, cpu=ps4, compiler=orbis, and glibc=orbis, does it _break_ if I specify all three flags? 
The error message is not very helpful saying "no supported toolchain for cpu=windows_x64, compiler=orbis, glibc=orbis" which indicates the `cpu` value is somehow replaced. The following options succeed, however: a) _only_ cpu=ps4 is specified, _OR_ b) host_cpu=ps4, compiler=orbis, glibc=orbis. Essentially, we ended up using option _a)_ which is identical to you `default_toolchain` suggestion.

It is absolutely possible that this is a misconfiguration of the CROSSTOOL, but if you have any comments regarding this, I'd be super grateful.

**Comment by aragos** [comment]
I believe this was fixed in #7151.
